File: Python/Single Number.py
"""
Given an array of integers, every element appears twice except for one. Find that single one.
Note:
Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
"""

class Solution:
    # @param A, a list of integer
    # @return an integer
    def singleNumber(self, A):
        r = 0
        for i in A:
            r ^= i
        return r

File: Java/Single Number.java
/*
Given an array of integers, every element appears twice except for one. Find that single one.
*/

public int singleNumber(int[] A) {
   	Set<Integer> set = new HashSet<>();
   	for (int x : A) {
      	if (set.contains(x)) {
         	set.remove(x);
      	} else {
         	set.add(x);
		} 
	}
   	return set.iterator().next();
}


public int singleNumber(int[] A) {
   	int num = 0;
   	for (int x : A)	{
	num ^= x; 
	}
	return num; 
}
